django/db/models/fields/json.py

```diff
@@ -0,0 +1,537 @@
+            if not lhs_only:
+                key_transforms.insert(0, previous.key_name)
+            previous = previous.lhs
+        lhs, params = compiler.compile(previous)
+        return (lhs, params, key_transforms) if not lhs_only else (lhs, params)
+
+
+class KeyTransform(PreprocessLhsMixin, Transform):
+    postgres_operator = '->'
+    postgres_nested_operator = '#>'
+
+    def __init__(self, key_name, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.key_name = key_name
+
+    def as_mysql(self, compiler, connection):
+        lhs, params, key_transforms = self.preprocess_lhs(compiler, connection)
+        json_path = compile_json_path(key_transforms)
+        if connection.vendor == 'mysql' and connection.mysql_is_mariadb:
+            return 'JSON_UNQUOTE(JSON_EXTRACT(%s, %%s))' % lhs, params + [json_path]
+        else:
+            return 'JSON_EXTRACT(%s, %%s)' % lhs, params + [json_path]
+
+    def as_oracle(self, compiler, connection):
+        lhs, params, key_transforms = self.preprocess_lhs(compiler, connection)
+        json_path = compile_json_path(key_transforms)
+        return "COALESCE(JSON_QUERY(%s, '%s'), JSON_VALUE(%s, '%s'))" % ((lhs, json_path) * 2), params
+
+    def as_postgresql(self, compiler, connection):
+        lhs, params, key_transforms = self.preprocess_lhs(compiler, connection)
+        if len(key_transforms) > 1:
+            return '(%s %s %%s)' % (lhs, self.postgres_nested_operator), params + [key_transforms]
+        try:
+            lookup = int(self.key_name)
+        except ValueError:
+            lookup = self.key_name
+        return '(%s %s %%s)' % (lhs, self.postgres_operator), params + [lookup]
+
+    def as_sqlite(self, compiler, connection):
+        return self.as_mysql(compiler, connection)
+
+
+class KeyTextTransform(KeyTransform):
+    postgres_operator = '->>'
+    postgres_nested_operator = '#>>'
+
+
+class KeyTransformTextLookupMixin:
+    """
+    Mixin for combining with a lookup expecting a text lhs from a JSONField
+    key lookup. On PostgreSQL, make use of the ->> operator instead of casting
+    key values to text and performing the lookup on the resulting representation.
+    On MySQL, JSON_UNQUOTE is applied to the lhs.
+    """
+    def __init__(self, key_transform, *args, **kwargs):
+        if not isinstance(key_transform, KeyTransform):
+            raise TypeError(
+                'Transform should be an instance of KeyTransform in order to use this lookup.'
+            )
+        key_text_transform = KeyTextTransform(
+            key_transform.key_name, *key_transform.source_expressions, **key_transform.extra
+        )
+        super().__init__(key_text_transform, *args, **kwargs)
+
+    def process_lhs(self, compiler, connection):
+        lhs, lhs_params = super().process_lhs(compiler, connection)
+        if connection.vendor == 'mysql' and not connection.mysql_is_mariadb:
+            return 'JSON_UNQUOTE(%s)' % lhs, lhs_params
```

Comment:
You use `JSON_UNQUOTE` for text transformations on MySQL and for all key transformation on MariaDB, this should be moved to the `mysql.operations.lookup_cast()`:
```python
    def lookup_cast(self, lookup_type, internal_type=None):
        lookup = '%s'
        if internal_type == 'JSONField':
            if self.connection.mysql_is_mariadb or lookup_type in (
                'iexact', 'contains', 'icontains', 'startswith', 'istartswith',
                'endswith', 'iendswith', 'regex', 'iregex',
            ):
                lookup = 'JSON_UNQUOTE(%s)'
        return lookup
```
after that `KeyTransformTextLookupMixin.process_lhs()` can be removed and `KeyTransform.as_mysql()` can be simplified:
```python
def as_mysql(self, compiler, connection):
    lhs, params, key_transforms = self.preprocess_lhs(compiler, connection)
    json_path = compile_json_path(key_transforms)
    return 'JSON_EXTRACT(%s, %%s)' % lhs, params + [json_path]
```